quixstreams/core/stream/stream.py

```diff
@@ -302,91 +299,101 @@ def diff(self, other: Self) -> Self:
             if the diff is empty, or pruning failed.
         :return: a new independent `Stream` instance whose root begins at the diff
         """
-        diff = self._diff_from_last_common_parent(other)
-
-        # The following operations enforce direct splitting.
-        # Enforcing direct splits relates to using one SDF to filter another.
-        # Specifically there are various unintuitive cases, especially when using a
-        # "split" SDF, where results will likely not be as expected, so we would
-        # rather raise an exception instead.
-        # See TestStreamingDataFrameSplitting test cases for examples.
-
-        # the easiest check that catches most issues: the "inner" (filtering) sdf
-        # should use same ref as the one being filtered; i.e. sdf[sdf.apply()].
-        if diff[0].pruned:
+        # Traverse the children of "self" and look for the "other" Stream among then
+        diff = []
+        self_found = False
+
+        for stream in other.root_path():
+            if stream is self:
+                self_found = True
+                break
+            # Assigning or filtering using branched SDFs
+            # may lead to an undefined behavior
+            if stream.is_branched():
+                raise InvalidOperation("Cannot assign or filter using a branched SDF")
+            elif stream.is_merged():
+                # Assigning or filtering using merged SDFs may lead
+                # to an undefined behavior too
+                raise InvalidOperation(
+                    "Cannot assign or filter using the SDF merged with another SDF"
+                )
+
+            diff.append(stream)
+
+        if not diff:
+            # The case of "sdf['x'] = sdf" or "sdf = sdf[sdf]"
+            raise InvalidOperation("Cannot assign or filter using the same SDF")
+
+        # Reverse the diff to be in "parent->children" order
+        diff.reverse()
+        diff_head = diff[0]
```

Comment:
As far as I can see `diff` is never used fully so to slightly improve readability and later avoid `diff[1:]` you could do this here.
```suggestion
        diff_head, *diff_rest = diff
```
or some other name like `diff_tail`, `diff_remainder`, idk.